### PR TITLE
Miscellaneous updates

### PR DESCRIPTION
**Pull request by amaanq**
I would like to see this parser integrated into nvim-treesitter, but I don't think it's ready for several reasons.

You should see the ci action I added, it will clone https://github.com/openscad/openscad and run tree-sitter parse on every .scad file in there. There are some failures in the screenshot below (note that some are intended to fail, you may ignore those but I am unsure which as I don't know the language)

<details>
<summary>Screenshot</summary>

image
</details>

In addition, I think you need to segment your expressions a bit more, so that we can easily have queries defined with `@parameter` and `@variable`, and maybe `@field, @attribute & @property` if they're applicable. Since I don't know OpenSCAD too well I made many assumptions with the queries.

Once you fix those problems and can make it easier to set certain queries, we can merge this into nvim-treesitter.

Thanks
-Amaan

### CODE REVIEW

**Comment by bollian** [comment]
Wow! This is an impressive amount of work. I'm going to start reading over it now.

File: grammar.js [hunk 1]
```diff
@@ -1,284 +1,391 @@
...
-  ),
-  if_clause: $ => prec.right(seq(
-    'if',
-    field('condition', $.parenthesized_expression),
-    field('consequence', $._comprehension_cell),
-    optional(
-      seq('else', field('alternative', $._comprehension_cell))
-    )
-  )),
+    list: ($) => brackets(seq(commaSep($._list_cell), optional(','))),
```
Comment (bollian) [inline grammar.js:309]: This `commaSep` invocation should probably now use the optional `trailing_separator` argument.

File: grammar.js [hunk 2]
```diff
@@ -1,284 +1,391 @@
...
-  ),
-  union_block: $ => curlies(repeat($._item)),
+    // modules
+    module_declaration: ($) => seq(
+      'module',
+      field('name', $.identifier),
+      field('parameters', $.parameters_declaration),
+      field('body', $._statement),
+    ),
+    parameters_declaration: ($) => parens(seq(commaSep($._parameter_declaration), optional(','))),
```
Comment (bollian) [inline grammar.js:171]: This `commaSep` invocation should probably now use the optional `trailing_separator` argument.

File: grammar.js [hunk 3]
```diff
@@ -1,284 +1,391 @@
...
+/**
+ * This callback should take a rule and places it in between a group (e.g. parentheses).
+ * @callback GroupingCallback
+ * @param {Rule} rule
+ *
+ * @return {SeqRule}
+ */
+
+/**
+ * Creates a rule that matches a binary operator in between two rules.
```
Comment (bollian) [inline grammar.js:60]: This looks like it was copy-pasted. It should read: "Creates a rule that accepts another rule with optional grouping symbols".

File: grammar.js [hunk 4]
```diff
@@ -1,284 +1,391 @@
...
+  return seq(field('left', rule), operator, field('right', rule));
 }
 
-const identifier_pattern = /[a-zA-Z_]\w*/;
+module.exports = grammar({
+  name: 'openscad',
 
-const rules = {
-  source_file: $ => repeat(choice($.use_statement, $._item)),
+  word: ($) => $.identifier,
```
Comment (bollian) [inline grammar.js:135]: I dislike the parentheses around the `$` here. It goes against the official tree-sitter examples, the repos owned by the tree-sitter github org, and it adds unnecessary visual noise.

**Comment by bollian** [comment]
To answer some of your questions: `@field`, `@attribute`, and `@property` do not apply to OpenSCAD.

So here are my initial thoughts:

- I don't want my CI status affected by future changes in the openscad repository. If they move the location of their examples, or they add new syntax that I don't parse yet, then my CI would preferably keep working. With that in mind, it's probably best to pin a particular revision, possibly using a git submodule
- I'm not familiar with ESLint, but I'd assume using just a `.eslintrc.json` file here would be best since the config is completely static

File: script/parse-examples [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  if [ ! -d "$repo" ]; then
+    git clone "https://github.com/$url" "$repo"
+  fi
+
+  pushd "$repo"
+  git fetch && git reset --hard "$sha"
+  popd
+}
+
+checkout examples/openscad openscad/openscad d7c5543dcadd7b60681305f9d1f8dc5de292f087
```
Comment (bollian) [inline script/parse-examples:19]: I'd prefer to just have a git submodule as part of the repository.

Comment (amaanq) [reply script/parse-examples:19]: Many main tree-sitter org parsers use a script like this. Git submodules can be sucky, and add unnecessary burden on a user testing to add it.

File: grammar.js (hunk 4 above)
Comment (amaanq) [reply grammar.js:135]: Yes, I forgot to disable that sorry

File: grammar.js (hunk 2 above)
Comment (amaanq) [reply grammar.js:171]: Fails with tests since (,) is valid

**Comment by amaanq** [comment]
Latest commit fixes parsing this piece of code:
```openscad
function f1(x, y) = echo("f1: ", x, y) 0.5 * x * x + 4 * y + 1;
```

**Comment by elentok** [comment]
Hi, thanks for doing this!! 🙏 🙏 🙏 

Any updates on this? really looking forward to having openscad treesitter support in Neovim :-)

**Comment by amaanq** [comment]
I can add my repo for the time being shortly till this PR is ironed out :)

**Comment by DeflateAwning** [comment]
Can you please merge this PR? It's been sitting here for a while, and it fixes an issue over at the VSCode support plugin.

**Comment by bollian** [comment]
Yes, I've definitely left this here too long. Part of the issue here is that the reformatting in this PR has resulted in a very difficult to read git diff. In order to actually understand the changes being made here, I need to roughly undo most of those changes.

I was actually considering trying to break this out into many smaller changesets and undoing the formatting changes. However, at the end of the day the functional changes are very good, and so long as this happens only _once_ I figure it's not so bad.

**Comment by DeflateAwning** [comment]
Thanks for merging this finally!

**Comment by amaanq** [comment]
Hey, I'd like to apologize about the unreadable diffs - I should have split them better among more commits, but I was pretty new to all the open source stuff and nvim-treesitter, really sorry about that, hopefully next time it'll be a lot better on my end :)

**Comment by bollian** [comment]
Since I figure this might affect people watching this issue, y'all should take a look at ce89936, since it comes with minor changes to numbers. My testing indicated that it didn't affect highlighting, which seems to be the usecase of most people in this PR.

Hey, no problem @amaanq. I really should've reached out to you sooner. 😅 Actually, could you explain for me the purpose behind the change to function_declaration in 9a05ed1? I know it's been a long time, but I can't think of a valid usecase for a sequence of expressions rather than a tree, and I don't see any tests that cover the change. I'm a little worried about pushing release 0.5 without understanding the change :).

**Comment by amaanq** [comment]
Hey, I'll help out managing the repo with some actions for checking for regressions + auto publishing tomorrow if that's all good with you, and I can revisit what I did then as well, I've forgotten by now :sweat_smile:

**Pull request merged**